File: ECE39500/exams/bak_midterm-2/midterm-2/PartA/StateFactory.py
from AddState import AddState
from CVCVState import CVCVState
from InitState import InitState
from EofState import EofState
from ErrorState import ErrorState
from VCVState import VCVState
import sys

class StateFactory():
    INT_MIN = -sys.maxsize - 1

    def __init__(self, compiler):
        self.compiler = compiler
        self.states = {
            "add": AddState(compiler),
            "CVCVState": CVCVState(compiler),
            "EofState": EofState(compiler),
            "ErrorState": ErrorState(compiler),
            "InitState": InitState(compiler),
            "VCVState": VCVState(compiler),
        }
        return
    
    def getState(self, state):
        return self.states[state]

File: ECE39500/exams/bak_midterm-2/PartA/StateFactory.java
import java.util.*;
import java.lang.reflect.*;

public class StateFactory {

   private Compiler compiler;
   private static String[ ] classes = {"AddState", "CVCVState", "EofState", 
                                       "ErrorState", "InitState", "VCVState"};

   // used to decode command names to the corresponding class name
   private static String[ ] commands = {"add", "CVCVState", "EofState", 
                                        "ErrorState", "InitState", "VCVState"};

   private static Map<String,CompilerState> states = 
      new HashMap<String,CompilerState>( );

   public StateFactory(Compiler compiler) {
      for (int i = 0; i < classes.length; i++ ) {

         Class<?> cls = null;
         Constructor<?> constructor = null;
         CompilerState state = null;

         try {
            cls = Class.forName(classes[i]);
         } catch (ClassNotFoundException e) {
            System.out.println("forName: "+classes[i]);
            e.printStackTrace( );
         }
         
         try {
            constructor = cls.getConstructor(Compiler.class);
         } catch (NoSuchMethodException e) {
            System.out.println("getConstructor: "+classes[i]);
            e.printStackTrace( );
         }

         try {
            state = (CompilerState) constructor.newInstance(compiler); 
            states.put(commands[i], state);
            if (Compiler.DEBUG) {
              System.out.println("Initializing state "+commands[i]+", status " + (state != null));
            }
         } catch (InstantiationException | IllegalAccessException | InvocationTargetException e) {
            System.out.println("newInstance: "+classes[i]);
            e.printStackTrace( );
         }
      }
   }

   public static CompilerState getState(String state) {
      CompilerState obj = states.get(state);
      if (obj == null) {
         System.out.println("Error getting state for class "+state); 
      }
      return obj;
   }
}
